[airvisualnode] Initial contribution for AirVisual Node Binding

3cky opened the pull request:
This binding integrates [AirVisual Node Air Quality Monitor](https://airvisual.com/node) with openHAB.

For complete information about this binding, please check [Readme.md](https://github.com/3cky/openhab2-addons/blob/airvisual-node-binding/addons/binding/org.openhab.binding.airvisualnode/README.md).

Signed-off-by: Victor Antonovich <v.antonovich@gmail.com>

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/about.html:
```diff
@@ -0,0 +1,32 @@
+<p>March 30, 2017</p>
+<h3>License</h3>
+
+<p>
```
I thing you need to add a `Third Party Content` section here regarding the license of the lib cifs in this file.

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,271 @@
+
+        schedulePoll();
+
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Node poll scheduled");
```
This will always hide the 2 previous configuration errors.

triller-telekom replied:
since you added the `return` statements above you will always end up with an offline thing which in reality is still in `INITIALIZING` state, I would suggest to remove this line and set the thing to ONLINE or OFFLINE inside your polling job.

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,271 @@
+
+        this.refreshInterval = config.refresh * 1000L;
+
+        schedulePoll();
```
Do you want to start the polling when there is a configuration error?

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    private synchronized void schedulePoll() {
+        stopPoll();
```
This is not needed as openHAB calls `dispose` before calling `initialize`. Thus the `stopPoll` is already been called.

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,271 @@
+    private synchronized void schedulePoll() {
+        stopPoll();
+        logger.debug("Scheduling poll for 500ms out, then every {} ms", refreshInterval);
+        pollFuture = scheduler.scheduleWithFixedDelay(new Runnable() {
```
Rewrite in java 8 notation: `scheduler.scheduleWithFixedDelay(this::poll(), ..`

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,271 @@
+            logger.debug("Could not connect to Node", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } catch (Exception e) {
+            logger.error("Unexpected error connecting to Node", e);
```
Don't log to error they are intended to indicate system wide failure.

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,271 @@
+        String url = "smb://" + nodeAddress +"/" + nodeShareName + "/" + NODE_JSON_FILE;
+        NtlmPasswordAuthentication auth = new NtlmPasswordAuthentication(null, nodeUsername, nodePassword);
+        try (SmbFileInputStream in = new SmbFileInputStream(new SmbFile(url, auth))) {
+            return IOUtils.toString(new SmbFileInputStream(new SmbFile(url, auth)), StandardCharsets.UTF_8.name());
```
You probably ment to use the variable `in` here that you moved to the line above.

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java:
```diff
@@ -0,0 +1,150 @@
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery");
+        cancelBackgroundDiscoveryFuture();
```
I assume stop is always called before start so there is probably no need to call this cancel?

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java:
```diff
@@ -0,0 +1,150 @@
+    @Override
+    protected void startScan() {
+        logger.debug("Starting scan");
+        scheduler.execute(scannerRunnable);
```
Maybe rewrite this as: `scheduler.execute(this::createScannerRunnable());` and remove the return part from `createScannerRunnable()`? (Although then the method name probably should be something more related to the running itself instead of creating the runnable`)

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java:
```diff
@@ -0,0 +1,150 @@
+                String workgroupUrl = "smb://" + AVISUAL_WORKGROUP_NAME +"/";
+                workgroupMembers = new SmbFile(workgroupUrl).listFiles();
+            } catch (Exception e) {
+                // Can't get workgroup member list
```
Are you sure it's not needed to log here? These exceptions will go undetected and might hide bugs.

Hilbrand commented on addons/binding/org.openhab.binding.airvisualnode/README.md:
```diff
@@ -0,0 +1,72 @@
+| status#used_memory           | Number    | Used memory                 |
+| status#timestamp             | DateTime  | Timestamp                   |
+
+## Example
```
This can also use an example of a Thing with parameter configuration.

3cky commented:
Thank you very much @Hilbrand for prompt PR reviewing, I'll fix that.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/AirVisualNodeBindingConstants.java:
```diff
@@ -0,0 +1,51 @@
+    public static final String CHANNEL_AQI_US = "aqi_us";
+    public static final String CHANNEL_PM_25 = "pm_25";
+    public static final String CHANNEL_TEMP_CELSIUS = "temp_celsius";
+    public static final String CHANNEL_TEMP_FAHRENHEIT = "temp_fahrenheit";
```
If you use UOM you only need one channel for temperature.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/AirVisualNodeBindingConstants.java:
```diff
@@ -0,0 +1,51 @@
+    public static final String CHANNEL_CO2_PPM = "co2_ppm";
+    public static final String CHANNEL_HUMIDITY = "humidity";
+    public static final String CHANNEL_AQI_CN = "aqi_cn";
+    public static final String CHANNEL_AQI_US = "aqi_us";
```
This is the same channel as the above one, but with a different unit, right? If you use UOM you only need one channel for this.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/AirVisualNodeBindingConstants.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
```
Please run `mvn license:format` to update all your headers to 2018.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/AirVisualNodeBindingConstants.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNull;
```
We are not using `@NonNull` but `@NonNullByDefault` on class level, see https://www.eclipse.org/smarthome/documentation/development/conventions.html

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,267 @@
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNull;
```
see above, please fix this for all files.

For now we also do not require that the whole code is annotated, so if you get stuck somewhere you can also leave those annotations out.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    private NodeData nodeData;
+
+    public AirVisualNodeHandler(@NonNull Thing thing) {
+        super(thing);
```
Please do not call `super` here because it sets your `thing` to ONLINE right way!

3cky replied:
Umm, looks like [BaseThingHandler constructor](https://github.com/eclipse/smarthome/blob/8d0856f424ae8115f1ca98d09c806a12a1827f47/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L112) doesn't sets `thing` to ONLINE mode, isn't it? Moreover, Eclipse complains about default constructor absence if I remove this `super` call.

triller-telekom replied:
Sorry for the confusion, I took the wrong exist O_o I was thinking of `initialize` which sets the `Thing` to ONLINE here https://github.com/eclipse/smarthome/blob/8d0856f424ae8115f1ca98d09c806a12a1827f47/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L242

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,267 @@
+        } catch (IOException e) {
+            logger.debug("Could not connect to Node", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } catch (Exception e) {
```
Please do not catch `Exception`, catch a more specific one instead. Because otherwise you might accidentally also catch `RuntimeExceptions` like `NullpointerExceptions`, etc.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,34 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-SymbolicName: org.openhab.binding.airvisualnode;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.2.0.qualifier
```
2.3.0.qualifier

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,34 @@
+ org.eclipse.smarthome.core.thing.binding.builder,
+ org.eclipse.smarthome.core.thing.type,
+ org.eclipse.smarthome.core.types,
+ org.osgi.service.component.annotations;resolution:=optional,
```
You can remove these now.

triller-telekom replied:
Please incorporate my comment and remove this line. It is not needed.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,145 @@
+				<context>network_address</context>
+				<label>Node Network Address</label>
+				<description>Node network address</description>
+				<required>true</required>
```
You do not need this node, you have set this already via `required="true"` on the parameter node.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,145 @@
+				<description>Node network address</description>
+				<required>true</required>
+			</parameter>
+			<parameter name="username" type="text" required="true">
```
If you provide a default you do not need the `required="true"`

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,145 @@
+
+		<!-- Channel groups -->
+
+		<channel-groups>
```
`channel-groups` are for making it possible to re-use a group of channels on several `thing-types` not to actually group them by same characteristics. so please use plain channels.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,145 @@
+			<channel id="co2_ppm" typeId="Co2_ppm" />
+			<channel id="humidity" typeId="Humidity" />
+			<channel id="aqi_cn" typeId="Aqi_cn" />
+			<channel id="aqi_us" typeId="Aqi_us" />
```
See my comment in the code, with UOM you will only need one of these 2 channels.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,145 @@
+			<channel id="aqi_us" typeId="Aqi_us" />
+			<channel id="pm_25" typeId="Pm_25" />
+			<channel id="temp_celsius" typeId="Temp_celsius" />
+			<channel id="temp_fahrenheit" typeId="Temp_fahrenheit" />
```
See my comment in the code, with UOM you will only need one of these 2 channels.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,145 @@
+	<channel-group-type id="Status">
+		<label>Status</label>
+		<channels>
+			<channel id="battery_level" typeId="system.battery-level" />
```
See my comment below, use a system channel for this.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,145 @@
+		<label>Status</label>
+		<channels>
+			<channel id="battery_level" typeId="system.battery-level" />
+			<channel id="wifi_strength" typeId="system.signal-strength" />
```
See my comment below, use a system channel for this.

3cky replied:
Hmm, could you clarify please, do you mean I should change these channel ids? Because both channels are already referenced to a system channel typeIds. Should I just rename channel ids to `battery-level` and `signal-strength`, or some more changes are required?

triller-telekom replied:
Upps, sorry apparently I stopped reading after `<channel id="battery_level"`... Sorry my fault.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,267 @@
+        Object value = null;
+
+        if (nodeData != null) {
+            String[] fullChannelIdFields = StringUtils.split(fullChannelId, "#");
```
I do not understand why you have to split the channelID by `#`. You should pass the correct ID to this method instead, splitting at `#` is not reliable.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java:
```diff
@@ -0,0 +1,144 @@
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery");
+        backgroundDiscoveryFuture = scheduler.scheduleWithFixedDelay(this::scan, 0, 300, TimeUnit.SECONDS);
```
Are you sure that the library you are using is doing the scan in the same way as windows clients do it? Because if you do a periodic scan every 5 minutes of the whole network your machine could be subject to be blocked by a net intrusion system in certain network setups. So I am unsure whether we should keep the automatic scan without user interaction in here.

3cky replied:
Well, the discovery process just requests SMB servers list from the local master browser using JCIFS library. Not sure how this operation will be treated by a net intrusion systems if repeated though. Do you recommend to disable periodic autodiscovery here?

triller-telekom replied:
I am no 100% sure. I just have in mind that we do not want discovery processes to "flood" the network without a user-interaction. A background scan works without a user interaction. If you would move it to `startScan` the user has to initiate the scan. But in your scenario is a corner case I guess and I am not a windows expert. Querying the master browser for the list of servers every 5min doesn't sound too bad though. Let's leave it in for now, I just wanted to understand what is actually happening here.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java:
```diff
@@ -0,0 +1,144 @@
+        try {
+            String workgroupUrl = "smb://" + AVISUAL_WORKGROUP_NAME +"/";
+            workgroupMembers = new SmbFile(workgroupUrl).listFiles();
+        } catch (Exception e) {
```
See above, please do not catch `Exception`

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java:
```diff
@@ -0,0 +1,144 @@
+                // Create discovery result
+                String nodeAddress = nodeNbtAddress.getInetAddress().getHostAddress();
+                DiscoveryResult result = DiscoveryResultBuilder.create(thingUID)
+                        .withProperty(AirVisualNodeConfig.ADDRESS, nodeAddress)
```
Since this is also unique for all devices, you should make it a `representationProperty`, see https://www.eclipse.org/smarthome/documentation/development/bindings/discovery-services.html

3cky commented:
Hello @triller-telekom, thank you for reviewing this PR!

Got some questions about applying UOM to the Air quality index (AQI). AQI US and AQI CN values are both dimensionless and differs only by method used for their computation (please see corresponding [Wikipedia article](https://en.wikipedia.org/wiki/Air_quality_index) if you're interested). Is the UOM still applicable in this case?

htreu commented:
Hi @3cky, regarding AQI I suggest not to use UOM in this case. AQI is a scale which maps to a combination of quantitative measurements. Its is itself not a quantitative property. You could use `ScaleTransformationService` to convert from one index to another.

3cky commented:
Hello @triller-telekom @htreu,

Thanks for your assistance! I've pushed updated PR, now UoM are used by all Node measurement channels. Not quite sure about the PM2.5 `Number:Density` channel though, I've used custom μg/m³ measurement unit https://github.com/openhab/openhab2-addons/pull/2805/files#diff-44b0bf6e12ae2c757dc0f1a4f2a50e0aR201 and it seems to work well in Items and Sitemaps, but not in Rules:

**avnode.rules**:
```
...
if (Livingroom_Pm25_Level.state > 25.0 [μg/m³]) {
    logInfo("avnode.rules", "PM2.5 level is above 25 μg/m³")
}
```
**openhab.log**:

```
2018-05-07 17:38:22.176 [WARN ] [el.core.internal.ModelRepositoryImpl] - Configuration model 'avnode.rules' has errors, therefore ignoring it: [40,42]: missing ')' at '['
[40,43]: no viable alternative at input 'μ'
[40,47]: no viable alternative at input '³'
```

Is this expected behaviour?

htreu commented:
@3cky unfortunately right now this is expected. The unit symbols `µ` and `³` are currently not supported by the script engine. Its missing in the DSL description. I will report back with a corresponding pull request in Eclipse SmartHome to fix this.

htreu commented:
@3cky once https://github.com/eclipse/smarthome/pull/5574 is merged and available for openHAB your scripts should work as expected.

3cky commented:
Hi @htreu,

Great! I'll wait for it, thank you!

htreu commented:
@3cky the corresponding PR already got merged and a new Eclipse SmartHome stable build also is available: DISCLAIMER: it introduced a new bug when triggering rules, so while µ and ^2/^3 are available it might nonetheless break your rule.

htreu commented:
Please note that unit support in scripts had to be refactored just recently. It now uses `|` instead of `[]` to add the unit. Like `20|°C`. Please try again and report back. Thanks.

3cky commented:
@htreu
 
I've tried this rule in openHAB 2.4.0-SNAPSHOT build 1299:

```
rule "AirVisual Node PM2.5 Level Rule"
when
    Item Livingroom_Pm25_Level changed
then
    if (Livingroom_Pm25_Level.state > 25.0|μg/m³) {
        logInfo("avnode.rules", "PM2.5 level is above 25 μg/m³")
    }
end
```
and I'm still getting the error:

> 2018-06-13 18:34:28.232 [WARN ] [el.core.internal.ModelRepositoryImpl] - Configuration model 'avnode.rules' has errors, therefore ignoring it: [40,44]: no viable alternative at input 'μ'
> [40,45]: missing ')' at 'g'
> [40,47]: no viable alternative at input 'm³'

3cky commented:
**Update**: I've changed `25.0|μg/m³` to `25.0|"μg/m³"` according to this documentation page https://www.openhab.org/docs/configuration/rules-dsl.html#number-item. Now 'no viable alternative at input' error is gone, but when `Livingroom_Pm25_Level` item value is changed I'm getting the next error:

> 2018-06-14 15:29:05.746 [ERROR] [ntime.internal.engine.RuleEngineImpl] - Rule 'AirVisual Node PM2.5 Level Rule': An error occurred during the script execution: μg not recognized (in 25.0 μg/m³ at index 5)

htreu commented:
I can for now confirm this issue. Testing with `1|"µm"` works fine, while `1|µg` produces the same error. Stay tuned.

htreu commented:
Here are my findings about this:

* The ESH internal tec.uom library does define a lot of `µ` prefixed units, also `µg`.
* There are two different characters `µ`:
  * The greek mu, unicode `\u03bc`
  * The micro sign, unicode `\u00b5`
* Only the micro sign is a valid character in terms of UoM.

Please try a script like `25.0|"\u00b5g"` which uses the afore mentioned micro sign. It worked out in my unit tests.

I'm not sure why and when a keyboard input produces which character. On my mac OS the combination `alt+m` always seems to produce the micro sign.

kaikreuzer commented:
Wow, not that is really nasty! @keilw, are you aware of this? What's the suggested UoM way to deal with it?

kaikreuzer commented:
@htreu Any explanation why you found `1|"µm"` working, though? What makes the difference?

keilw commented:
Have to check, but I know when I created JUnit tests for the temperature question, it occurred to me again, there are 2 ways to render °C, one with 2 separate characters, while there is an Unicode char that combines both ;-O

kaikreuzer commented:
We know, we just got rid off that "feature"... https://github.com/eclipse/smarthome/issues/5669

htreu commented:
When I author the unit test with my setup everything works as expected. After @3cky's last report I did copy his script from the comment and reproduced the error. All `µ` characters I type on my keyboard seem to be valid micro signs.
If you look closely in my comment (https://github.com/openhab/openhab2-addons/pull/2805#issuecomment-397321886) you will see a difference in rendering of both `µ` chars.

keilw commented:
There are other formatters/parsers, especially in the UCUM bundle, but that would have to be used via CQ.

keilw commented:
I could not find a Unicode `\u03bc` in uom-se, or do you mean the plain text character used in `MetricPrefix.MICRO`?

3cky commented:
Oops, such a difference between 'µ' and 'μ' is quite unexpected for me. :) Honestly speaking, I just copied 'GREEK SMALL LETTER MU"' character from GNOME Character Map utility. Now I changed it to "U+00B5 MICRO SIGN" and rule is working just fine! Thanks for your efforts in explaining such weird rule behaviour. 
BTW, it seem that these characters are undistinguishable for browsers and text editors, when I'm doing "Ctrl+F" for any of them both characters are marked as matched.

htreu commented:
Thanks @3cky for clarification. @keilw I guess this is not a real issue then since copying the "wrong" character from a symbol table is not considered default behaviour.
@triller-telekom I guess the review can continue then.

keilw commented:
Glad it works that way. I checked the symbols in uom-se and the next API (where `MICRO` becomes available in an implementation-neutral way) and even the plain text MICRO SIGN should be "U+00B5" in all places.

3cky commented:
Thank you everyone, I've updated rules according to the UoM changes and bumped version to 2.4.0-SNAPSHOT. Please let me know if any other changes are required.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -52,80 +58,46 @@
+		<item-type>Number:Dimensionless</item-type>
+		<label>AQI</label>
 		<description>Air Quality Index (US)</description>
+		<category>Flow</category>
```
I am not sure if this is the right category. What is flowing here? Better leave it out.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/ESH-INF/thing/thing-types.xml:
```diff
@@ -52,80 +58,46 @@
 		<label>PM2.5</label>
 		<description>PM2.5 level, μg/m&#179;</description>
-		<state readOnly="true" pattern="%.1f μg/m&#179;" />
+		<category>Flow</category>
```
I am not sure if this is the right category. What is flowing here? Better leave it out.

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/README.md:
```diff
@@ -73,9 +65,68 @@ airvisualnode:avnode:1a2b3c4 [ address="192.168.1.32", username="airvisual", pas
+Number:Dimensionless Livingroom_Humidity "Humidity [%d %unit%]" <humidity> {channel="airvisualnode:avnode:1a2b3c4:humidity"}
+Number:Dimensionless Livingroom_CO2_Level "CO₂" <flow> {channel="airvisualnode:avnode:1a2b3c4:co2"}
+Number:Dimensionless Livingroom_Aqi_Level "Air Quality Index" <flow> { channel="airvisualnode:avnode:1a2b3c4:aqi" }
+Number:Density Livingroom_Pm25_Level  "PM2.5 Level" <line> { channel="airvisualnode:avnode:1a2b3c4:pm_25" }
```
Where do you have the category "line" from? I cannot find it in our documentation: https://www.eclipse.org/smarthome/documentation/concepts/categories.html

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/README.md:
```diff
@@ -73,9 +65,68 @@ airvisualnode:avnode:1a2b3c4 [ address="192.168.1.32", username="airvisual", pas
+Number:Dimensionless Livingroom_CO2_Level "CO₂" <flow> {channel="airvisualnode:avnode:1a2b3c4:co2"}
+Number:Dimensionless Livingroom_Aqi_Level "Air Quality Index" <flow> { channel="airvisualnode:avnode:1a2b3c4:aqi" }
+Number:Density Livingroom_Pm25_Level  "PM2.5 Level" <line> { channel="airvisualnode:avnode:1a2b3c4:pm_25" }
+DateTime Livingroom_Aqi_Timestamp "AQI Timestamp [%1$tH:%1$tM]" <clock> { channel="airvisualnode:avnode:1a2b3c4:timestamp" }
```
Where do you have the category "clock" from? I cannot find it in our documentation: https://www.eclipse.org/smarthome/documentation/concepts/categories.html

triller-telekom commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/AirVisualNodeBindingConstants.java:
```diff
@@ -8,44 +8,46 @@
+    public static final String CHANNEL_TEMP_CELSIUS = "temperature";
     public static final String CHANNEL_BATTERY_LEVEL = "battery_level";
-    public static final String CHANNEL_WIFI_STRENGTH = "wifi_strength";
+    public static final String CHANNEL_WIFI_STRENGTH = "signal_strength";
```
Since you now use the system channels, its probably better to use their definitions directly, i.e.: `DefaultSystemChannelTypeProvider.SYSTEM_SYSTEM_CHANNEL_SIGNAL_STRENGTH.getUID().getId()`

Same applies to other system channels of course.

3cky commented:
Hello @triller-telekom,

Thank you for reviewing this PR and please sorry for my late response. I removed problematic item categories, cleaned up MANIFEST.MF from all non-essential import packages and did changes in system channels handling code. Please let me know if there are any other fixes needed.

3cky commented:
@triller-telekom do you mean this line? `org.osgi.service.component.annotations;resolution:=optional,`
When I remove it, Eclipse reports the error below:

![image](https://user-images.githubusercontent.com/894050/43130606-b54517a8-8f3f-11e8-9074-4947b6c36152.png)

Is this normal or I did miss something?

kaikreuzer commented:
@3cky Are you using the official openHAB IDE? If not, you will need to enable the DS annotation generator, I just commented this on https://github.com/eclipse/smarthome/issues/5954#issuecomment-407351591.

3cky commented:
@kaikreuzer you're right, I'm using not openHAB IDE but regular Eclipse IDE installation. I followed your instructions and error is gone now. Thanks! So I'll remove annotations package import from the manifest file.

martinvw commented on addons/binding/org.openhab.binding.airvisualnode/.project:
```diff
@@ -0,0 +1,39 @@
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
```
This should not be here?

martinvw commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/AirVisualNodeBindingConstants.java:
```diff
@@ -0,0 +1,56 @@
+            .getUID().getId();
+
+    // List of all supported Thing UIDs
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections
```
Afair, `Set`'s should be annotated to have a nullable generic, because otherwise the compiler might assume that a set cannot return null but it can.

@maggu2810 or am I mistaken now?

triller-telekom replied:
@martinvw You are right, but where is that `@Nullable` annotation that you are referring to?

martinvw replied:
Missing and the class is annotated to be not Null

triller-telekom replied:
Ah now I understand my misunderstanding of your comment...

`Set`s should **not** be annotated to have a `@Nullable` generic, because that would express that we allow to add `null` to the collection.

The problem you are mentioning is that a `get(item)` on a `Set` might always return `null` if the requested `item` is not in the `Set`. But that should be covered by external annotations which annotate (in this case) the JDT. But we do not have them, both PRs https://github.com/eclipse/smarthome/pull/4513 and https://github.com/eclipse/smarthome/pull/4217 were closed.

Long story short: The line here is correct.

maggu2810 replied:
Nothing to add to @triller-telekom's explanation. :wink:

martinvw replied:
Great thanks guys!

martinvw commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,234 @@
+
+    public AirVisualNodeHandler(Thing thing) {
+        super(thing);
+        gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES).create();
```
👍

martinvw commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    private String getNodeJsonData() throws IOException {
+        String url = "smb://" + nodeAddress +"/" + nodeShareName + "/" + NODE_JSON_FILE;
```
Very small formatting error, if in doubt format all your (Java) files

martinvw commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/handler/AirVisualNodeHandler.java:
```diff
@@ -0,0 +1,234 @@
+                case CHANNEL_TIMESTAMP:
+                    // It seem the Node timestamp is Unix timestamp converted from UTC time plus timezone offset.
+                    // Not sure about DST though, but it's best guess at now
+                    Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
```
We have more modern options since Java 8, I would prefer if you use them instead of the calendar

martinvw commented on addons/binding/org.openhab.binding.airvisualnode/src/main/java/org/openhab/binding/airvisualnode/internal/discovery/AirVisualNodeDiscoveryService.java:
```diff
@@ -0,0 +1,145 @@
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery");
+        backgroundDiscoveryFuture = scheduler.scheduleWithFixedDelay(this::scan, 0, 300, TimeUnit.SECONDS);
```
Why not specify the much more explicit

```java
..., 0, 5, TimeUnit.MINUTES);

3cky commented:
Hello @martinvw, thank you for your comments! I did some fixes, please check them out.

kaikreuzer commented:
@3cky Please ping me once you have pushed an update - thanks!

3cky commented:
Hello @kaikreuzer, thank you for your patience. I made fixes in accordance with your comments, please check.

3cky commented:
Great, thank you all guys for your assistance!
